### PR TITLE
Remove `TypeVariableOriginKind` and `ConstVariableOriginKind`

### PR DESCRIPTION
**Pull request by compiler-errors**
It's annoying to have to import `TypeVariableOriginKind` just to fill it with `MiscVariable` for almost every use. Every other usage other than `TypeParameterDefinition` wasn't even used -- I can see how it may have been useful once for debugging, but I do quite a lot of typeck debugging and I've never really needed it.

So let's just remove it, and keep around the only useful thing which is the `DefId` of the param for `var_for_def`.

This is based on #123006, which removed the special use of `TypeVariableOriginKind::OpaqueInference`, which I'm pretty sure I was the one that added.

r? lcnr or re-roll to types

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/pat.rs [hunk 1]
```diff
@@ -1344,7 +1344,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             self.next_ty_var(
                 // FIXME: `MiscVariable` for now -- obtaining the span and name information
```
Comment (lcnr) [inline compiler/rustc_hir_typeck/src/pat.rs:1345]: please grep for `MiscVariable`

File: compiler/rustc_infer/src/infer/opaque_types/mod.rs [hunk 2]
```diff
@@ -72,10 +72,7 @@ impl<'tcx> InferCtxt<'tcx> {
                     // FIXME(compiler-errors): We probably should add a new TypeVariableOriginKind
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/opaque_types/mod.rs:72]: outdated

File: compiler/rustc_middle/src/infer/unify_key.rs [hunk 3]
```diff
@@ -106,16 +105,9 @@ impl ToType for ty::FloatVarValue {
...
     pub span: Span,
-}
-
-/// Reasons to create a const inference variable
-#[derive(Copy, Clone, Debug)]
-pub enum ConstVariableOriginKind {
-    MiscVariable,
-    ConstInference,
-    ConstParameterDefinition(Symbol, DefId),
+    // `DefId` of the const parameter this was instantiated for, if any.
```
Comment (lcnr) [inline compiler/rustc_middle/src/infer/unify_key.rs:109]:
```suggestion
    /// `DefId` of the const parameter this was instantiated for, if any.
    ///
    /// This should only be used for diagnostics.
```

**Review by lcnr — COMMENTED** [review]
please make sure to update all relevant comments

after that r=me, a really nice cleanup imo ❤️

**Comment by matthiaskrgr** [comment]
mh may need another rebase

**Pull request merged**